[Kkenjji] iP

Kkenjji opened the pull request:
# 👾  Clank: Chatbot x Task Manager

“A plan is a list of actions arranged in whatever sequence is thought likely to achieve an objective.” – **John Argenti**

Clank is a intelligent **task management chatbot** that helps you stay organized and productive. 🧠  

## ✨ Features  
- **Text-based** interaction 📝  
- **Easy to learn** 📖  
- **Blazing fast** ⚡  
- **Completely FREE!** 🎉  

## 🛠 How to Use  
1. **Download** the latest version from [here](https://github.com/kkenjji/ip/releases).  
2. **Double-click** to run the application or run it using the command:
```
java -jar clank.jar
```
3. **Add your tasks** via simple text commands.  
4. Let Clank do the rest! 🤖  

## ✅ Upcoming Features  
- [x] Basic task management  
- [x] GUI support  
- [x] Reminder notifications 

## 🔧 Code Example  
Here's how a task is added by **Clank**:  

```java
public void addTask(Task task) {
    tasks.add(task);
}
```

---

🚀 **Stay productive with Clank!**

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Kkenjji committed:
Configure .gitignore to ignore .class files

Kkenjji committed:
Rename Duke to Clank

Kkenjji committed:
Add echoing functionality

Kkenjji committed:
Add ability to store and list out texts

Kkenjji committed:
Add mark/unmark feature

Kkenjji committed:
Add program exit for "bye"

Kkenjji committed:
Add TaskManager to handle different tasks

Kkenjji committed:
Bug fixes

Kkenjji committed:
Add program exit

Kkenjji committed:
Add exceptions to handle invalid inputs

Kkenjji committed:
Add delete task feature

Kkenjji committed:
Add test cases for Clank

Kkenjji committed:
Rename constants

nicholasohjj commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+
+            try {
+                switch (first) {
+                case "bye":
+                    System.out.println(byeMessage);
+                    System.exit(0);
+                    break;
```
Avoid Using `System.exit(0)` in a Loop-Based Application

It forces an abrupt termination instead of allowing a graceful exit.

```suggestion
                case "bye":
                    System.out.println(byeMessage);
                    break; // Breaks out of the while loop, terminating normally
```

Kkenjji replied:
Fixed!

nicholasohjj commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+        String input;
+
+        while (true) {
+            input = scanner.nextLine();
+            input = input.toLowerCase();
```
Avoid `toLowerCase()` on Entire Input

This can cause unintended issues when parsing user commands that involve case-sensitive data.

```suggestion
            input = scanner.nextLine();
            String[] splitted = input.split(" ", 2);
            String first = splitted[0].toLowerCase(); // Convert only the command
            String second = (splitted.length == 2) ? splitted[1] : "";
```

Kkenjji replied:
Fixed!

nicholasohjj commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+        String byeMessage = " Bye. Hope to see you again soon!\n";
+        System.out.println(welcomeMessage);
+
+        Scanner scanner = new Scanner(System.in);
```
The `Scanner `object is never closed, leading to a potential resource leak.

Use a `try-with-resources` statement to ensure `scanner` is automatically closed.

```suggestion
        try (Scanner scanner = new Scanner(System.in)) {
    while (true) {
        String input = scanner.nextLine();
        // (rest of the loop remains unchanged)
    }
}
```

Kkenjji replied:
Fixed!

nicholasohjj commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+            System.out.println(line);
+
+            try {
+                switch (first) {
```
Improve Readability of `switch` Statements

The `switch` cases handle different user commands but some cases call methods directly, while others throw exceptions. This inconsistency makes it harder to read.

Use a helper method like `processCommand()` to separate command handling logic from input reading.

Kkenjji replied:
I have extracted all different commands out as classes of their own, thanks for the advice!

nicholasohjj commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+            } catch (ClankException e) {
+                System.out.println(e.getMessage());
+            } catch (Exception e) {
+                System.out.println("An unexpected error occured: " + e.getMessage());
```
Improve Exception Handling for Unexpected Errors

The catch block for general exceptions prints only the error message, which might not be useful for debugging.

```suggestion
                System.out.println("An unexpected error occured: " + e.getMessage());
                e.printStackTrace(); // Prints the full stack trace

```

Kkenjji replied:
Will change accordingly, thanks!

nicholasohjj reviewed (changes requested):
nits but they enhance code structure, maintainability, and reliability.

LGTM after these refinements! 👍

ShiJianXi commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+        String line = "____________________________________________________________\n";
+        String welcomeMessage = " Greetings! I'm Clank, Ratchet's best friend.\n" +
+                " I heard you are in trouble, and I'm here to help.\n" +
+                " How may I assist you?\n" +
```
Might have some minor coding standards violation here, refer to the [coding standard guide](https://se-education.org/guides/conventions/java/intermediate.html), on the section on Layout.
Should have breaks before an operators.

Kkenjji replied:
Thanks for the reminder!

ShiJianXi commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+        String welcomeMessage = " Greetings! I'm Clank, Ratchet's best friend.\n" +
+                " I heard you are in trouble, and I'm here to help.\n" +
+                " How may I assist you?\n" +
+                line;
```
```suggestion
        String welcomeMessage = " Greetings! I'm Clank, Ratchet's best friend.\n" 
                     + " I heard you are in trouble, and I'm here to help.\n" 
                     + " How may I assist you?\n" 
                     + line;
```

Kkenjji replied:
I see, I've forgotten about that. Thanks for the reminder!

ShiJianXi commented on src/main/java/Clank.java:
```diff
@@ -0,0 +1,69 @@
+            input = input.toLowerCase();
+
+            String[] splitted = input.split(" ", 2);
+            String first = splitted[0];
+            String second = (splitted.length == 2) ? splitted[1] : "";
```
Perhaps a more meaningful variable name can be used instead of first and second.

Kkenjji replied:
Thanks for the advice. I have changed the variable names.

ShiJianXi reviewed (commented):
LGTM. Code is clean, just some nits on coding standards.

Kkenjji committed:
Add save feature

Kkenjji committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save feature
  Rename constants

Kkenjji committed:
Add save message

Kkenjji committed:
Refine code logic

Kkenjji committed:
Add ability to understand inputs with time and date

Kkenjji committed:
Rename list method in TaskManager class

Kkenjji committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Rename list method in TaskManager class
  Add ability to understand inputs with time and date
  Refine code logic

Kkenjji committed:
Refactor code to extract more classes

Kkenjji committed:
Add addCommand class

Kkenjji committed:
Add addTask method in TaskList

Kkenjji committed:
Extract multiple classes, Parser, Ui, Storage, TaskList, and classes

Kkenjji committed:
Extract code into multiple classes

Kkenjji committed:
Merge branch 'branch-MoreOOP'

* branch-MoreOOP:
  Extract code into multiple classes
  Extract multiple classes, Parser, Ui, Storage, TaskList, and classes
  Add addTask method in TaskList
  Add addCommand class
  Refactor code to extract more classes

Kkenjji committed:
Package all classes

Kkenjji committed:
Merge branch 'branch-Packages'

* branch-Packages:
  Package all classes

Kkenjji committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

Kkenjji committed:
Set mainClass to Clank

Kkenjji committed:
Add checkstyle files

Kkenjji committed:
Merge branch 'branch-Gradle'

* branch-Gradle:
  Add checkstyle files
  Set mainClass to Clank

shelialia commented on src/main/java/clank/command/AddCommand.java:
```diff
@@ -0,0 +1,73 @@
+        String[] parts = input.split(" ", 2);
+
+        String mainCommand = parts[0].toLowerCase();
+        String[] details = parts.length > 1 ? parts[1].split("/by | /from | /to ", 3) : new String[]{""};
```
Could this variable name be more specific?

Kkenjji replied:
I'll change it to taskDetails, more specific.

shelialia commented on src/main/java/clank/command/AddCommand.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) {
```
Could this method name be more specific? "Execute" is rather vague, a reader may find difficulty understanding the purpose of the method. Perhaps we could rename it to a specific action such as "addAndStoreTask"

Kkenjji replied:
Oh, this method overrides the execute method in its parent class (Command.java).

shelialia commented on src/main/java/clank/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+        }
+
+        try {
+            String type = parts[0];
```
"type" is rather vague as well, and is frequently used in coding concepts. perhaps we could use "taskType" instead?

Kkenjji replied:
Thanks for the recommendation. Will modify accordingly.

shelialia commented on src/main/java/clank/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+            String description = parts[2];
+
+            Task task = null;
+            switch (type) {
```
Good job on following the coding standard indentation for the switch-case statement!

shelialia commented on src/main/java/clank/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
```
Good job on ensuring the boolean sounds like a boolean.

shelialia reviewed (commented):
LGTM. Several occurrences of vagueness in variable names, but overall clean code 👍

Kkenjji committed:
Add test cases for AddCommand, Deadline, Event, Todo, Parser

Kkenjji committed:
Merge branch 'branch-JUnit'

* branch-JUnit:
  Add test cases for AddCommand, Deadline, Event, Todo, Parser

Kkenjji committed:
Add print messages for all commands

Kkenjji committed:
Add JAR build configurations

Kkenjji committed:
Merge branch 'branch-Jar'

* branch-Jar:
  Add JAR build configurations
  Add print messages for all commands

Kkenjji committed:
Add find command

Kkenjji committed:
Fix style errors

Kkenjji committed:
Add Javadocs for Clank.java

Kkenjji committed:
Add Javadocs for classes in clank.utility

Kkenjji committed:
Add Javadocs for Todo, Deadline, and Event

Kkenjji committed:
Add Javadocs for Task and TaskList

Kkenjji committed:
Add 'throws' keyword for methods in Storage.java

Kkenjji committed:
Add 'throws' keyword for parse method in Parser.java

Kkenjji committed:
Add Javadocs for custom exceptions in clank.exception package

Kkenjji committed:
Add Javadocs for all classes in clank.command package

Kkenjji committed:
Add Javadocs for test files

Kkenjji committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fix style errors

Kkenjji committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find command

# Conflicts:
#	src/main/java/clank/utility/Parser.java

Kkenjji committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add Javadocs for test files
  Add Javadocs for all classes in clank.command package
  Add Javadocs for custom exceptions in clank.exception package
  Add 'throws' keyword for parse method in Parser.java
  Add 'throws' keyword for methods in Storage.java
  Add Javadocs for Task and TaskList
  Add Javadocs for Todo, Deadline, and Event
  Add Javadocs for classes in clank.utility
  Add Javadocs for Clank.java

# Conflicts:
#	src/main/java/clank/command/Command.java
#	src/main/java/clank/task/Deadline.java
#	src/main/java/clank/task/Event.java
#	src/main/java/clank/task/TaskList.java
#	src/main/java/clank/utility/Storage.java
#	src/main/java/clank/utility/Ui.java
#	src/test/java/clank/command/AddCommandTest.java
#	src/test/java/clank/task/DeadlineTest.java
#	src/test/java/clank/task/EventTest.java
#	src/test/java/clank/task/TodoTest.java
#	src/test/java/clank/utility/ParserTest.java

Kkenjji committed:
Add Javadocs for find command

Kkenjji committed:
Reorder operators to align with coding conventions

Kkenjji committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Reorder operators to align with coding conventions

Kkenjji committed:
Edit welcome message

Kkenjji committed:
Add close method to close scanner in Ui.java

Kkenjji committed:
Modify findTasks to accept multiple keywords using varargs

Kkenjji committed:
Merge branch 'branch-A-Varargs'

* branch-A-Varargs:
  Modify findTasks to accept multiple keywords using varargs

Kkenjji committed:
Rename some variables

Kkenjji committed:
Update build.gradle to support javafx

Kkenjji committed:
Fix FindCommand bug

Kkenjji committed:
Add GUI

Kkenjji committed:
Add Javadocs

Kkenjji committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add Javadocs
  Add GUI
  Fix FindCommand bug
  Update build.gradle to support javafx

Kkenjji committed:
Fix setClank

Kkenjji committed:
Refactor Command.java

Kkenjji committed:
Refactor MainWindow.java

Refactoring the class improves readability and modularity.

Extracting helper methods for handling user input and bot responses
makes the code cleaner and more readable.

Changes:
- Extract `displayResponse()` to handle user and bot responses.
- Extract `scheduleExit()` to handle program exiting logic.

These improvements enhance readability by improving code organization.

Kkenjji committed:
Add Javadocs for displayResponse() and
scheduleExit() in MainWindow.java

Kkenjji committed:
Refactor Task.java

Refactoring the class improves readability and modularity.

Extracting helper methods for task creation to make the code
cleaner and more readable.

Changes:
- Extract `createTask()` method to simplify `fromSavedFormat()`.

These improvements enhance readability by improving code organization.

Kkenjji committed:
Refactor TaskList.java

The `TaskList` class contained redundant logic and inefficient loops.

Refactoring the class improves readability.

Changes:
- Extracted `validateIndex()` method to remove duplicate
    index validation logic.

This change enhance code organization, making the class more readable.

Kkenjji committed:
Refactor AddCommand.java

The `AddCommand` constructor contained repetitive logic and
deep nesting.

Refactoring the class improves readability while preserving
functionality.

Changes:
- Extract `createTodo()`, `createDeadline()`, and `createEvent()`
methods to remove redundant code.
- Add Javadocs for all three newly created methods.

These changes enhance maintainability by improving code organization.

Kkenjji committed:
Use Java Streams for findTasks() and listTasks()

Kkenjji committed:
Use Java Streams in showMatchingTasks()

Kkenjji committed:
Add assertions in all command classes, Storage.java and Ui.java

Kkenjji committed:
Merge pull request #1 from Kkenjji/branch-A-CodeQuality

Branch (Code Quality)

Kkenjji committed:
Merge branch 'master' of https://github.com/Kkenjji/ip

* 'master' of https://github.com/Kkenjji/ip:
  Refactor AddCommand.java
  Refactor TaskList.java
  Refactor Task.java
  Add Javadocs for displayResponse() and scheduleExit() in MainWindow.java
  Refactor MainWindow.java
  Refactor Command.java

Kkenjji committed:
Merge branch 'master' into branch-A-Streams

* master:
  Refactor AddCommand.java
  Refactor TaskList.java
  Refactor Task.java
  Add Javadocs for displayResponse() and scheduleExit() in MainWindow.java
  Refactor MainWindow.java
  Refactor Command.java

Kkenjji committed:
Merge branch 'branch-A-Streams' into branch-A-Assertions

* branch-A-Streams:
  Use Java Streams in showMatchingTasks()
  Use Java Streams for findTasks() and listTasks()
  Refactor AddCommand.java
  Refactor TaskList.java
  Refactor Task.java
  Add Javadocs for displayResponse() and scheduleExit() in MainWindow.java
  Refactor MainWindow.java
  Refactor Command.java

# Conflicts:
#	src/main/java/clank/command/AddCommand.java

Kkenjji committed:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add assertions in all command classes, Storage.java and Ui.java
  Use Java Streams in showMatchingTasks()
  Use Java Streams for findTasks() and listTasks()

Kkenjji committed:
Add Java CI workflow

Kkenjji committed:
Fix unable to print message bug when adding task

Kkenjji committed:
Fix code style violations

Kkenjji committed:
Capitalize `am` and `pm` in test methods in
DeadlineTest and EventTest classes

Kkenjji committed:
Fix file paths in runtest.sh and runtest.bat to
correctly compile and run

Kkenjji committed:
Remove IO redirection tests

Kkenjji committed:
Add reminder feature to notify users of upcoming tasks

Implemented a reminder feature that allows users to check for
upcoming tasks within a specified number of days.

Changes:
- Created `ReminderCommand` to fetch and display upcoming tasks.
- Modified `Ui` to format and display reminders in a structured format.
- Updated `Clank` to automatically trigger reminders at startup.
- Users can manually check reminders using the `reminder <days>` command.

This feature ensures that users are aware of pending deadlines and
events without needing to manually check their task list.

Kkenjji committed:
Add autosave after deletion

Kkenjji committed:
Merge branch 'branch-B-Reminders'

* branch-B-Reminders:
  Add autosave after deletion
  Add reminder feature to notify users of upcoming tasks

Kkenjji committed:
Add Javadocs to ReminderCommand.java

Kkenjji committed:
Remove unused variable

Kkenjji committed:
Update MainWindow.java to close the application immediately

Kkenjji committed:
Improve Clank GUI

Kkenjji committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Improve Clank GUI
  Update MainWindow.java to close the application immediately

Kkenjji committed:
Add some unique messages

Kkenjji committed:
Merge branch 'branch-A-Personality'

* branch-A-Personality:
  Add some unique messages

Kkenjji committed:
Add test cases for delete, find, mark, reminder, unmark commands

Kkenjji committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  Add test cases for delete, find, mark, reminder, unmark commands

Kkenjji committed:
Use enum types to take care of exceptions in ClankException.java

Kkenjji committed:
Change deletion message

Kkenjji committed:
Add sample UI screenshot

Kkenjji committed:
Fix coding standard violations

Kkenjji committed:
Fix load task bug

Kkenjji committed:
Add README and updated sample Ui screenshot

Kkenjji committed:
Fix typo

Kkenjji committed:
Fix typo

Kkenjji committed:
Configure DateTimeFormatters to use upper case

Kkenjji committed:
Rewrite filePath to the correct path

Kkenjji committed:
Rename clank.jar to Clank.jar

Kkenjji committed:
Remove failed to parse exception

Kkenjji committed:
Add AI.md to declare use of AI

Kkenjji committed:
Minor changes to README.md
